packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+// Type-only imports
+import {TemplateResult, DirectiveResult} from 'lit-html';
+
+import {
+  nothing,
+  noChange,
+  Directive,
+  Part,
+  NODE_PART,
+  AttributePart,
+} from 'lit-html';
+
+import {$private} from 'lit-html/private-ssr-support.js';
+
+const {getTemplateHtml, marker, markerMatch, boundAttributeSuffix} = $private;
+
+import {digestForTemplateResult} from 'lit-html/hydrate.js';
+
+import {ElementRenderer} from './element-renderer.js';
+
+import {escapeAttribute, escapeTextContent} from './util/escaping.js';
+
+// types only
+import {DefaultTreeDocumentFragment} from 'parse5';
+
+import {
+  traverse,
+  parseFragment,
+  isCommentNode,
+  isElement,
+} from './util/parse5-utils.js';
+
+import {isRenderLightDirective} from 'lit-html/directives/render-light.js';
+import {LitElement} from 'lit-element';
+import {LitElementRenderer} from './lit-element-renderer.js';
+import {reflectedAttributeName} from './reflected-attributes.js';
+
+declare module 'parse5' {
+  interface DefaultTreeElement {
+    isDefinedCustomElement?: boolean;
+  }
+}
+
+Directive.prototype.resolve = function (
+  this: Directive,
+  _part: Part,
+  values: unknown[]
+) {
+  return this.render(...values);
+};
+
+const templateCache = new Map<
+  TemplateStringsArray,
+  {
+    ops: Array<Op>;
+    attrNames: Array<string>;
+  }
+>();
+
+/**
+ * Operation to output static text
+ */
+type TextOp = {
+  type: 'text';
+  value: string;
+};
+
+/**
+ * Operation to output dynamic text from the associated template result value
+ */
+type NodePartOp = {
+  type: 'node-part';
+  index: number;
+  useCustomElementInstance?: boolean;
+};
+
+/**
+ * Operation to output an attribute with bindings. Includes all bindings for an
+ * attribute, like an attribute template part or AttributeComitter.
+ */
+type AttributePartOp = {
+  type: 'attribute-part';
+  index: number;
+  name: string;
+  strings: Array<string>;
+  tagName: string;
+  useCustomElementInstance?: boolean;
+};
+
+/**
+ * Operator to create a custom element instance.
+ */
+type CustomElementOpenOp = {
+  type: 'custom-element-open';
+  tagName: string;
+  ctor: {new (): HTMLElement};
+  staticAttributes: Map<string, string>;
+};
+
+/**
+ * Operation to render a custom element's attributes. This is separate from
+ * `custom-element-open` because attribute/property parts go in between and need
+ * to run and be set on the instance before we render the element's final
+ * attributes.
+ */
+type CustomElementAttributesOp = {
+  type: 'custom-element-attributes';
+};
+
+/**
+ * Operation to render a custom element's children, usually its shadow root.
+ */
+type CustomElementChildrenOp = {
```

Comment:
I can't recall. I think I had been referring to these type of children as just children because with renderRoot they might not be in the shadow root. This is simpler to think about if we remove `createRenderRoot` :)

Reply:
Since I moved creating `<template shadowroot>` to this op and renamed the renderer API to `renderShadow`, I'll go ahead and rename it.

Reply:
Although now SSR won't be compatible with `createRenderRoot() { return this }` like you said (I think it would basically work before). Do we care?